### PR TITLE
flutter: Bump and add dart cache to flutter

### LINKED ISSUE #98405: Should dart and cache be made available when installing Flutter?
**Describe the bug**
I have installed flutter using Nix, everything related to flutter seems to be working great. Having some issues though when it comes to editor integration (I'm trying to set up Emacs with lsp-dart) . The problem seems to be that lsp-dart wants to use the dart version that comes with the flutter installation. Usually it can be found in `$FLUTTER_ROOT/bin/dart` and `$FLUTTER_ROOT/bin/cache`. I have found a solution where i edited `flutter.nix` to create symlinks to flutter unwrapped dir, where dart and cache is available. This doesn't integrate well with lsp-dart, but by setting the flutter sdk directory manually to the unwrapped dir in `/nix/store` I got it working. This will however require me to update my Emacs config whenever nix is changed.

I considered writing an overlay to take care of this, but as I'm pretty new to both nix and flutter I couldn't figure out where to begin. Therefore I decided to ask, how do you deal with dart and dart_analysis_server? Should dart and cache directory be made available somehow? 

Any help would be greatly appreciated! 

**Notify maintainers**
@babariviere 
@ericdallo 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.0-47-generic, Ubuntu, 20.04.1 LTS (Focal Fossa)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(storvik): `"home-manager, nixpkgs-21.03pre243111.51428e8d382"`
 - nixpkgs: `/home/storvik/.nix-defexpr/channels/nixpkgs`

### PR DESCRIPTION
**Pull request by ericdallo**
###### Motivation for this change
* Flutter requires dart-sdk cache to work propertly
* bump flutter version to latest

Fixes #98405

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/flutter/default.nix [hunk 1]
```diff
@@ -10,25 +12,28 @@ in {
...
   beta = mkFlutter rec {
-    pname = "flutter-beta";
-    channel = "beta";
-    version = "1.20.0-7.2.pre";
-    filename = "flutter_linux_${version}-${channel}.tar.xz";
-    sha256Hash = "0w89ig5vi4spa95mf08r4vvwni7bzzdlyhvr9sy1a35qmf7j9s6f";
-    patches = getPatches ./patches/beta;
+    pname = "flutter";
+    channel = "stable";
+    version = "1.22.0";
```
Comment (babariviere) [inline pkgs/development/compilers/flutter/default.nix:24]: Shouldn't it be `beta` instead of `stable`? Latest `beta` version is `1.22.0-12.3.pre`

Comment (ericdallo) [reply pkgs/development/compilers/flutter/default.nix]: Sorry, For some reason when I run `flutter --version` from beta channel it returns exactly the same for `stable`. But bu the git version of flutter repo, I think you are right, done!

Comment (ericdallo) [reply pkgs/development/compilers/flutter/default.nix]: Could you please take a look again @babariviere ?

**Review by babariviere — APPROVED** [review]
LGTM

**Review by thiagokokada — APPROVED** [review]

**Comment by ericdallo** [comment]
@danieldk or @worldofpeace could you please merge?

**Comment by storvik** [comment]
Just tested this with Nix and Ubuntu. Flutter and dart development tools seems to be working as expected :+1:

**Comment by ericdallo** [comment]
@jonringer maybe you can help us merging that?

**Comment by ericdallo** [comment]
@marsam ?

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -8635,7 +8635,11 @@ in
 
   flutterPackages =
     recurseIntoAttrs (callPackage ../development/compilers/flutter { });
-  flutter = flutterPackages.stable;
+  flutter = {
+    stable = flutterPackages.stable;
+    beta = flutterPackages.beta;
+    dev = flutterPackages.dev;
+  };
```
Comment (marsam) [inline pkgs/top-level/all-packages.nix:8642]: usually we don't define new attrsets in `all-packages.nix`, would you mind applying the following?
```suggestion
  flutter = flutterPackages.stable;
  flutterBeta = flutterPackages.beta;
  flutterDev = flutterPackages.dev;
```
besides that, LGTM

Comment (ericdallo) [reply pkgs/top-level/all-packages.nix]: Thanks! Done @marsam

**Comment by ericdallo** [comment]
Thank you @marsam :heart:

**Pull request merged**